typed_python/compiler/expression_conversion_context.py

```diff
@@ -777,28 +781,32 @@ def convert_expression_ast(self, ast):
             return pythonObjectRepresentation(self, ast.s)
 
         if ast.matches.BoolOp:
-            def convertBoolOp(depth=0):
+            with self.subcontext():
+                return_type = OneOfWrapper.mergeTypes((self.convert_expression_ast(v).expr_type.typeRepresentation for v in ast.values))
+
+            def convertBoolOp(return_type, depth=0):
                 with self.subcontext() as sc:
                     value = self.convert_expression_ast(ast.values[depth])
-                    value = TypedExpression.asBool(value)
-                    if value is not None:
+                    bool_value = TypedExpression.asBool(value)
+                    value_expr = value.convert_to_type(return_type).nonref_expr
+                    if bool_value is not None:
                         if depth == len(ast.values) - 1:
-                            sc.expr = value.nonref_expr
+                            sc.expr = value_expr
                         else:
-                            tail_expr = convertBoolOp(depth + 1)
+                            tail_expr = convertBoolOp(return_type, depth + 1)
 
                             if ast.op.matches.And:
                                 sc.expr = native_ast.Expression.Branch(
-                                    cond=value.nonref_expr, true=tail_expr, false=native_ast.falseExpr)
+                                    cond=bool_value.nonref_expr, true=tail_expr, false=value_expr)
                             elif ast.op.matches.Or:
                                 sc.expr = native_ast.Expression.Branch(
-                                    cond=value.nonref_expr, true=native_ast.trueExpr, false=tail_expr)
+                                    cond=bool_value.nonref_expr, true=value_expr, false=tail_expr)
                             else:
                                 raise Exception(f"Unknown kind of Boolean operator: {ast.op.Name}")
 
                 return sc.result
 
-            return TypedExpression(self, convertBoolOp(), typeWrapper(bool), False)
+            return TypedExpression(self, convertBoolOp(return_type), typeWrapper(return_type), False)
```

Comment:
this is the source of the problem. the 'branch' statements above are selecting between different versions of 'value_expr', which are themselves reference types. But then we create a TypedExpression (whose value is a non-reference) out of the merge of the different pathways. this isn't supposed to happen, because we need OneOf's to be reference to be able to manipulate them correctly.

Reply:
Thanks, will add more tests.

Reply:
There's a problem with the test I committed.  Working on it.